WorkloadManagementSystem/Utilities/PilotCStoJSONSynchronizer.py

```diff
@@ -355,32 +333,36 @@ def _upload(self, pilotDict=None, filename='', pilotScript=''):
     # http://docs.python-requests.org/en/master/user/advanced/#post-multiple-multipart-encoded-files
     # But well, maybe too much optimization :-)
 
-    if pilotDict:  # this is for the pilot.json file
-      if not self.pilotFileServer:
-        self.log.warn("NOT uploading the pilot JSON file, just printing it out")
+    if not self.pilotFileServer:
+      self.log.warn("No pilotFileServer, nowhere to upload")
+      if pilotDict:
         print(json.dumps(pilotDict, indent=4, sort_keys=True))  # just print here as formatting is important
-        return S_OK()
-
-      data = {'filename': self.jsonFile, 'data': json.dumps(pilotDict)}
 
-    else:  # we assume the method is asked to upload the pilots scripts
-      if not self.pilotFileServer:
-        self.log.warn("NOT uploading", filename)
-        return S_OK()
-
-      # ALWAYS open binary when sending a file
-      with open(pilotScript, "rb") as psf:
-        script = psf.read()
-      data = {'filename': filename, 'data': script}
-
-    for pfServer in self.pilotFileServer.replace(' ', '').split(','):
-      resp = requests.post('https://%s/DIRAC/upload' % pfServer,
-                           data=data,
-                           verify=self.casLocation,
-                           cert=self.certAndKeyLocation)
-
-      if resp.status_code != 200:
-        return S_ERROR(resp.text)
-
-    self.log.info('-- File and scripts upload done --')
-    return S_OK()
+    else:
+      if pilotDict:  # this is for the pilot.json file
+        filename = self.jsonFile
+        script = json.dumps(pilotDict)
+        with open(filename, 'w') as jf:
+          json.dump(script, jf)
+
+      else:  # we assume the method is asked to upload the pilots scripts
+        # ALWAYS open binary when sending a file
+        with open(pilotScript, "rb") as psf:
+          script = psf.read()
+
+      for pfServer in self.pilotFileServer.replace(' ', '').split(','):
+        try:
+          data = {'filename': filename, 'data': script}
+          resp = requests.post('https://%s/DIRAC/upload' % pfServer,
+                               data=data,
+                               verify=self.casLocation,
+                               cert=self.certAndKeyLocation)
+          # This POST works iff the webserver implements the upload function
+          # Which is done in WebAppDIRAC.WebApp.handler.RooHandler.RootHandler.web_upload()
+          # So, this location is specific to DIRAC WebApp.
+          if resp.status_code != 200:
+            self.log.error("Status code != 200", "%s returned when POSTing on %s" % (resp.text, pfServer))
+
+        except requests.ConnectionError:
+          # if we are here is probably because we are not trying to upload to a WS that does not have expose a POST API
```

Comment:
Too many "not"s? 
```suggestion
          # if we are here it is probably because we are trying to upload to a WS that does not expose a POST API
```
WS=Webserver?

You also have to update the function docstring return value